### DISCUSSION TITLE
Global CLI

### DISCUSSION THREAD

**Discussion by erhise** [erhise/offline-blitz#20] [Ideas]
If proposal is to make offline-blitz global with `npm i -g offline-blitz` what are all thinkable scenarios to handle?

1st scenario: Your own sandbox repo
```
My-own-sandbox/
  app.ts
  example.ts
  custom-pipe.ts
```
You run `offline-blitz app.ts`?

Should there be some default file as it would be enough to run `offline-blitz`?

2nd scenario: Inside project
```
Big-repo/
  subfolder/
     library.component.ts
     quick-test.ts
```
You run `offline-blitz subfolder/quick-test.ts`?

3rd scenario: Having internal dependencies
```
//app.ts
import { mySimpleFunction } from './other';

const result = mySimpleFunction(33) * 2;
console.log(result);

//other.ts
export function mySimpleFunction(x: number) {
   return x - 3;
}
```
You run `offline-blitz app.ts`

4rd scenario: Empty dir
```
//simple-test.ts
import { of } from 'rxjs';

of('works').subscribe(v => console.log(v));
```
You run `offline-blitz simple-test.ts`

Should it work?

@siguse

**Comment by erhise** [comment]
5rd scenario: Internal dependencies
```
my-repo/
  node_modules/
     some-lib
  quick-test.ts

//quick-test.ts

import { calc } from 'some-lib'

const test = calc(2+2);
console.log(test);
```
You run `offline-blitz quick-test.ts`

**Comment by lurande-faran** [comment]
Good scenarios. This is not an all easy question. I think we could have some default filenames that will be run if you do `offline-blitz folder/`. In that case we can look for app.ts, index.ts or bin.ts, or something like that, within the folder? And in this case we watch the whole folder for changes. In the scenario when you do `offline-blitz quick-test.ts` we should only run and watch that file. But if that file, as in scenario 3, has dependencies, we would want to watch that file for changes as well. We should look into how to properly extract that information in a good way. I know chokidar has the abillity to add files to the watcher during runtime. So as long as we can detect the dependecy, we should not have a problem adding these files to the watcher as a result of a change->restart. But in scenario 5 the dependecy is from node_modules. In that case we should not watch it. We could probably use tsconfig to determine this. What do you think?

**Reply by lurande-faran** [reply]
TLDR
I think all scenarios should work, but it's a question of how to set up the watcher in the different scenarios. @erhise

**Reply by erhise** [reply]
Yes, it also interesting on how to properly handle error
1. Determine what context (what files to watch, what file to run, what type of file it is, how that file look like)
2. Handle supported scenarios but throw error for not supported cases (which somehow needs to be determine in 1)
3. Run file with watch (if supported in 2 => how to integrate custom tsconfig if ts file)

**Reply by lurande-faran** [reply]
3. The ts-node compiler is automatically picking up the tsconfig file in the working directory. So if we set cwd properly in the compiler, we can access these properties on the compiler object (this is currently happening in the offline-blitz directory 🙂).

**Reply by lurande-faran** [reply]
1. and 2. which are the unsupported cases? 🙂

**Reply by erhise** [reply]
I was just thinking out loud. But I what I had in mind was firstly the scenario where you are trying to run some file (mistakenly or not) with a whole lot of dependencies which should result in fail? ¯\_(ツ)_/¯ . The other scenario I was thinking about was how to integrate the terminal interface in a good public API. Maybe run `offline-blitz --interactive` / `offline-blitz -i` and specify global objects that is picked up. This would then result in error e.g. 'actions are missing, did you intend to use it?'

**Reply by lurande-faran** [reply]
A lot of dependencies is fine by me actually, as long as we only watch the relevant paths for changes. 🙂 Like anything excluded in tsconfig or maybe files located outside of the cwd should be ignored by the watcher? 
Yes, an -i flag and/or --interactive flag would be nice. Or is it enough to do something like `import { buildUI } from 'offline-blitz';`? Or is that too obscure?

**Reply by erhise** [reply]
I think this would depend if the user install the package globally or locally?

Global: ~~`import { buildUI } form 'offline-blitz`~~ since it normally resolve from local env (dont think we should mess with that)
Local: can't run `offline-blitz -i`

**Reply by lurande-faran** [reply]
Ah, ok! Did not know that 😁 Makes sense!
